Purchases/SwiftSources/LocalReceiptParsing/ReceiptParser.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  ReceiptParser.swift
+//  Purchases
+//
+//  Created by Andrés Boedo on 7/22/20.
+//  Copyright © 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+
+class ReceiptParser {
+    private let objectIdentifierParser: ASN1ObjectIdentifierBuilder
+    private let containerBuilder: ASN1ContainerBuilder
+    private let receiptBuilder: AppleReceiptBuilder
+
+    init(objectIdentifierParser: ASN1ObjectIdentifierBuilder = ASN1ObjectIdentifierBuilder(),
+         containerBuilder: ASN1ContainerBuilder = ASN1ContainerBuilder(),
+         receiptBuilder: AppleReceiptBuilder = AppleReceiptBuilder()) {
+        self.objectIdentifierParser = objectIdentifierParser
+        self.containerBuilder = containerBuilder
+        self.receiptBuilder = receiptBuilder
+    }
+
+    func parse(from receiptData: Data) throws -> AppleReceipt {
+        let intData = [UInt8](receiptData)
+
+        let asn1Container = try containerBuilder.build(fromPayload: ArraySlice(intData))
+        guard let receiptASN1Container = try findASN1Container(withObjectId: ASN1ObjectIdentifier.data,
+                                                               inContainer: asn1Container) else {
+            throw ReceiptReadingError.dataObjectIdentifierMissing
+        }
+        let receipt = try receiptBuilder.build(fromContainer: receiptASN1Container)
+        return receipt
+    }
+}
+
+private extension ReceiptParser {
+    func findASN1Container(withObjectId objectId: ASN1ObjectIdentifier,
+                           inContainer container: ASN1Container) throws -> ASN1Container? {
+        if container.encodingType == .constructed {
+            for (index, internalContainer) in container.internalContainers.enumerated() {
+                if internalContainer.containerIdentifier == .objectIdentifier {
+                    let objectIdentifier = objectIdentifierParser.build(fromPayload: internalContainer.internalPayload)
+                    if objectIdentifier == objectId && index < container.internalContainers.count - 1 {
```

Comment:
Can you clarify the `index < container.internalContainers.count - 1` for me? I don't think I get it 🤔

Reply:
it's just to make sure that there _is_ a next container available after the object identifier.